CombinatorialGames/Mathlib/Concept.lean

```diff
@@ -0,0 +1,412 @@
+@[simp]
+theorem upperPolar_iUnion (f : ι → Set α) :
+    upperPolar r (⋃ i, f i) = ⋂ i, upperPolar r (f i) :=
+  (gc_upperPolar_lowerPolar r).l_iSup
+
+@[simp]
+theorem lowerPolar_iUnion (f : ι → Set β) :
+    lowerPolar r (⋃ i, f i) = ⋂ i, lowerPolar r (f i) :=
+  upperPolar_iUnion ..
+
+theorem upperPolar_iUnion₂ (f : ∀ i, κ i → Set α) :
+    upperPolar r (⋃ (i) (j), f i j) = ⋂ (i) (j), upperPolar r (f i j) :=
+  (gc_upperPolar_lowerPolar r).l_iSup₂
+
+theorem lowerPolar_iUnion₂ (f : ∀ i, κ i → Set β) :
+    lowerPolar r (⋃ (i) (j), f i j) = ⋂ (i) (j), lowerPolar r (f i j) :=
+  upperPolar_iUnion₂ ..
+
+theorem subset_lowerPolar_upperPolar (s : Set α) :
+    s ⊆ lowerPolar r (upperPolar r s) :=
+  (gc_upperPolar_lowerPolar r).le_u_l _
+
+theorem subset_upperPolar_lowerPolar (t : Set β) :
+    t ⊆ upperPolar r (lowerPolar r t) :=
+  subset_lowerPolar_upperPolar _ t
+
+@[simp]
+theorem upperPolar_lowerPolar_upperPolar (s : Set α) :
+    upperPolar r (lowerPolar r <| upperPolar r s) = upperPolar r s :=
+  (gc_upperPolar_lowerPolar r).l_u_l_eq_l _
+
+@[simp]
+theorem lowerPolar_upperPolar_lowerPolar (t : Set β) :
+    lowerPolar r (upperPolar r <| lowerPolar r t) = lowerPolar r t :=
+  upperPolar_lowerPolar_upperPolar _ t
+
+theorem upperPolar_anti : Antitone (upperPolar r) :=
+  (gc_upperPolar_lowerPolar r).monotone_l
+
+theorem lowerPolar_anti : Antitone (lowerPolar r) :=
+  upperPolar_anti _
+
+/-! ### Concepts -/
+
+variable (α β)
+
+/-- The formal concepts of a relation. A concept of `r : α → β → Prop` is a pair of sets `s`, `t`
+such that `s` is the set of all elements that are `r`-related to all of `t` and `t` is the set of
+all elements that are `r`-related to all of `s`. -/
+structure Concept where
+  /-- The extent of a concept. -/
+  extent : Set α
+  /-- The intent of a concept. -/
+  intent : Set β
+  /-- The intent consists of all elements related to all elements of the extent. -/
+  upperPolar_extent : upperPolar r extent = intent
+  /-- The extent consists of all elements related to all elements of the intent. -/
+  lowerPolar_intent : lowerPolar r intent = extent
+
+namespace Concept
+
+variable {r r' α β}
+variable {c d : Concept α β r} {c' : Concept α α r'}
+
+attribute [simp] upperPolar_extent lowerPolar_intent
+
+@[ext]
+theorem ext (h : c.extent = d.extent) : c = d := by
+  obtain ⟨s₁, t₁, h₁, _⟩ := c
+  obtain ⟨s₂, t₂, h₂, _⟩ := d
+  dsimp at h₁ h₂ h
+  substs h h₁ h₂
+  rfl
+
+theorem ext' (h : c.intent = d.intent) : c = d := by
+  obtain ⟨s₁, t₁, _, h₁⟩ := c
+  obtain ⟨s₂, t₂, _, h₂⟩ := d
+  dsimp at h₁ h₂ h
+  substs h h₁ h₂
+  rfl
+
+theorem extent_injective : Injective (@extent α β r) := fun _ _ => ext
+
+theorem intent_injective : Injective (@intent α β r) := fun _ _ => ext'
+
+/-- Copy a concept, adjusting definitional equalities. -/
+def copy (c : Concept α β r) (e : Set α) (he : e = c.extent) (i : Set β) (hi : i = c.intent) :
+    Concept α β r where
+  extent := e
+  intent := i
+  upperPolar_extent := by simp_all
```

Comment:
(In a proof, I would suggest replacing this with `rw [he, hi, upperPolar_extent]`. Do we prefer `simp_all` here since it's in `def`?)